Subject: Enhance php-config script

Changes:

* Add --php-api, --zend-extension-api and --zend-module-api options

  This adds new options to the php-config script to output the PHP
  internal API version number (PHP_API_VERSION in main/php.h), the Zend
  module API number (ZEND_MODULE_API_NO in Zend/zend_modules.h) and Zend
  extension API number (ZEND_EXTENSION_API_NO in
  Zend/zend_extensions.h). The man pages are updated with brief
  information about these values. There are existing PHP installations,
  that already use similar patches, such as --phpapi on Debian and
  Ubuntu, where the output value matches the ZEND_MODULE_API_NO.

* Remove sed from php-config script

  During the configuration step, the Autoconf configure options are
  replaced in main/build-defs.h.in and scripts/php-config.in. Instead of
  doing redundant quotes removal step each time the php-config script is
  run by the user on installed PHP, they can be removed once before the
  variable substitution step directly at the PHP configuration phase
  when building PHP. This also removes the unused CONFIGURE_OPTIONS
  Make variable in root and phpize generated Makefiles. The configure
  options in main/build-defs.h.in remain quoted as before.

  Also, sed substituted value can have issues in cases where build is
  happening on one machine and targeted machine has different path to
  the sed command. For example, on build machine it is at /usr/bin/sed,
  and on target at /bin/sed.

  This should be further improved because some PHP installations
  manually adjust the php-config script and inject variables in
  php-config script, such as CFLAGS. For example, variables passed to
  configure after quotes removal can include malformed value with spaces
  and without any quotes around them. Perhaps removing quotes should be
  done only on configure options and not variable arguments.
---
 build/php.m4                 |  3 ++-
 configure.ac                 |  6 +++++
 scripts/man1/php-config.1.in | 18 ++++++++++++++
 scripts/php-config.in        | 48 +++++++++++++++++++++---------------
 4 files changed, 54 insertions(+), 21 deletions(-)

diff --git a/build/php.m4 b/build/php.m4
index e45b22b766..74f0e25b01 100644
--- a/build/php.m4
+++ b/build/php.m4
@@ -2105,7 +2105,8 @@ EOF
   echo '"[$]@"' >> $1
   chmod +x $1
   CONFIGURE_COMMAND="$CONFIGURE_COMMAND $CONFIGURE_OPTIONS"
-  AC_SUBST([CONFIGURE_COMMAND])
+  dnl Remove quotes.
+  CONFIGURE_OPTIONS=$(echo $CONFIGURE_OPTIONS | $SED -e "s#'##g")
   AC_SUBST([CONFIGURE_OPTIONS])
 ])
 
diff --git a/configure.ac b/configure.ac
index 56c22f1a80..68de471911 100644
--- a/configure.ac
+++ b/configure.ac
@@ -1363,6 +1363,9 @@ exec_prefix=$old_exec_prefix
 libdir=$old_libdir
 datadir=$old_datadir
 
+ZEND_EXTENSION_API_NO=$($EGREP '#define ZEND_EXTENSION_API_NO' $srcdir/Zend/zend_extensions.h | "${SED}" -e 's/^[^0-9]*//')
+PHP_API_VERSION=$($EGREP '#define PHP_API_VERSION ' $srcdir/main/php.h | "${SED}" -e 's/^[^0-9]*//')
+
 AC_SUBST([INCLUDE_PATH])
 AC_SUBST([EXPANDED_PEAR_INSTALLDIR])
 AC_SUBST([EXPANDED_EXTENSION_DIR])
@@ -1381,6 +1384,9 @@ AC_SUBST([SAPI_LIBNAME_STATIC])
 AC_SUBST([PHP_VERSION])
 AC_SUBST([PHP_VERSION_ID])
 AC_SUBST([PHP_LDFLAGS])
+AC_SUBST([PHP_API_VERSION])
+AC_SUBST([ZEND_MODULE_API_NO])
+AC_SUBST([ZEND_EXTENSION_API_NO])
 
 PHP_UTILIZE_RPATHS
 
diff --git a/scripts/man1/php-config.1.in b/scripts/man1/php-config.1.in
index 9944ac91d7..d0c2b60bf8 100644
--- a/scripts/man1/php-config.1.in
+++ b/scripts/man1/php-config.1.in
@@ -73,6 +73,24 @@ PHP version
 .PD 0
 .B \-\-vernum
 PHP version as integer
+.TP
+.PD 0
+.B \-\-php-api
+Internal PHP API version number
+.TP
+.PD 0
+.B \-\-zend-module-api
+Internal API version number for PHP extensions (dynamically loaded with
+\fBextension\fP INI directive). This is the \fBZEND_MODULE_API_NO\fP number in
+Zend/zend_modules.h and ensures that built extension is compatible with
+particular PHP build.
+.TP
+.PD 0
+.B \-\-zend-extension-api
+Internal API version number for Zend extensions in PHP (dynamically loaded with
+\fBzend_extension\fP INI directive), such as opcache, debuggers, profilers, etc.
+This is the \fBZEND_EXTENSION_API_NO\fP number in Zend/zend_extensions.h and
+ensures that built extension is compatible with particular PHP build.
 .RS
 .PD 1
 .P
diff --git a/scripts/php-config.in b/scripts/php-config.in
index 87c20089bb..fc5372d3e0 100644
--- a/scripts/php-config.in
+++ b/scripts/php-config.in
@@ -1,6 +1,5 @@
 #! /bin/sh
 
-SED="@SED@"
 prefix="@prefix@"
 datarootdir="@datarootdir@"
 exec_prefix="@exec_prefix@"
@@ -23,6 +22,9 @@ php_sapis="@PHP_INSTALLED_SAPIS@"
 ini_dir="@EXPANDED_PHP_CONFIG_FILE_SCAN_DIR@"
 ini_path="@EXPANDED_PHP_CONFIG_FILE_PATH@"
 php_embed_type="@PHP_EMBED_TYPE@"
+php_api="@PHP_API_VERSION@"
+zend_module_api="@ZEND_MODULE_API_NO@"
+zend_extension_api="@ZEND_EXTENSION_API_NO@"
 
 # Set php_cli_binary and php_cgi_binary if available
 for sapi in $php_sapis; do
@@ -50,9 +52,6 @@ elif test "$php_embed_type" = "static"; then
   php_embed_lib=@SAPI_LIBNAME_STATIC@
 fi
 
-# Remove quotes
-configure_options=`echo $configure_options | $SED -e "s#'##g"`
-
 case "$1" in
 --prefix)
   echo $prefix;;
@@ -86,26 +85,35 @@ case "$1" in
   echo $version;;
 --vernum)
   echo $vernum;;
+--php-api)
+  echo $php_api;;
+--zend-module-api)
+  echo $zend_module_api;;
+--zend-extension-api)
+  echo $zend_extension_api;;
 *)
   cat << EOF
 Usage: $0 [OPTION]
 Options:
-  --prefix            [$prefix]
-  --includes          [$includes]
-  --ldflags           [$ldflags]
-  --libs              [$libs]
-  --extension-dir     [$extension_dir]
-  --include-dir       [$include_dir]
-  --lib-dir           [$lib_dir]
-  --lib-embed         [$php_embed_lib]
-  --man-dir           [$man_dir]
-  --php-binary        [$php_binary]
-  --php-sapis         [$php_sapis]
-  --ini-path          [$ini_path]
-  --ini-dir           [$ini_dir]
-  --configure-options [$configure_options]
-  --version           [$version]
-  --vernum            [$vernum]
+  --prefix             [$prefix]
+  --includes           [$includes]
+  --ldflags            [$ldflags]
+  --libs               [$libs]
+  --extension-dir      [$extension_dir]
+  --include-dir        [$include_dir]
+  --lib-dir            [$lib_dir]
+  --lib-embed          [$php_embed_lib]
+  --man-dir            [$man_dir]
+  --php-binary         [$php_binary]
+  --php-sapis          [$php_sapis]
+  --ini-path           [$ini_path]
+  --ini-dir            [$ini_dir]
+  --configure-options  [$configure_options]
+  --version            [$version]
+  --vernum             [$vernum]
+  --php-api            [$php_api]
+  --zend-module-api    [$zend_module_api]
+  --zend-extension-api [$zend_extension_api]
 EOF
   exit 1;;
 esac
